### DISCUSSION TITLE
MirthConnect 3.8.0 has no indexed timestamp fields

### DISCUSSION THREAD

**Discussion by BuddhaBuddy1** [nextgenhealthcare/connect#5366] [Q&A] [answered]
Got message bombed by a client today, so I'm trying to make some queries to monitor the rate of messages.

The DB structure seem strange, creating different tables for different channels. Not sure if that was an installation option, but since I'm actually only interested in a single channel, that's fine. Probably a performance optimization?

The real question is why there are no indexed timestamp fields? We use MySQL, and looking in the statistics table, there isn't a single indexed timestamp field anywhere in the schema. That seems strange, since the UI has timestamp filters. Did our install fail, or is that intended? It seems like m_{channel_id}.received_date,  mm_{channel_id}.received_date,  mm_{channel_id}.send_date,  and mm_{channel_id}.response_date really could use one each.

Before I started slapping indexes everywhere, I just wanted to see if that was intentional because it caused performance issues or something. Seems like it would be a performance issue to NOT have them, but index behavior on large tables can be surprising. Even with date filters, the UI is super slow on searches unless the date range is tiny, though, so I'm wondering if this is why.

**Comment by jonbartels** [comment, marked as answer]
Ok so let me take these in reverse order. For context I've been using MC since like 2006 but I've never been and am not current a dev nor acting in any official capacity for anyone. I trust my memory of the history of the project but I expect I'm missing some details.

You're not the first person to ask this but I'm not sure what to search for here or on https://forums.mirthproject.io for a canonical answer. I know this has come up before and that there were even suggestions to partition the MC tables, but AFAIK none have been implemented to-date (pun intended!)

> I just wanted to see if that was intentional because it caused performance issues or something

Roughly yes. MC is INSERT and UPDATE intensive as it receives and processes messages. The reads at runtime are almost all by message ID. So having indexes is a bigger performance hit for how MC talks to its DB. Historically it was done for performance and size considerations. In my opinion, on a modern DB with a minimal amount of monitoring and tuning *AND* with MC set up to prune regularly this isn't a problem. If you have a large channel history and an untuned DB, searches over long history will suuuck.

If your DB is tuned and you're pruning, putting indexes in place seems reasonable. Definitely read your DB doco for index types and options to make date searches efficient. 

> Did our install fail, or is that intended?

No your install didn't fail. The lack of indexes is expected. 

> The DB structure seem strange, creating different tables for different channels. Not sure if that was an installation option, but since I'm actually only interested in a single channel, that's fine. Probably a performance optimization?

Also yes. Mirth connect creates a set of tables per channel. This lets the messages from channel 1 be separate from channel 2 in different tables. You can see the structure by downloading this HTML file. https://gist.github.com/jonbartels/e1f9805209c8c34feb1ff86cd651f783 

This goes wayy back to at least 3.0 and probably into the 2.0 series. Its an intentional design choice. The table names are based off of the LOCAL channel ID (an int) and you can correlate that to the channel name or ID with the `d_channels` table.  You can see an example of the dynamic SQL to report on this at https://gist.github.com/MichaelLeeHobbs/67980d165fc68880eb2ab283c673244b (a MySQL port would be awesome and useful if you're able to contribute!).

**Reply by BuddhaBuddy1** [reply]
Thanks for the details, very helpful!

I get the schema, and figured it was to have each channel partitioned into its own set of tables.

The things I'm trying to do are pretty different than what the PG script does. I'm just trying to see who is the latest client to "clean up" their DB and trigger a ton of old messages into my system. The data would also be useful to show a live-ish histogram of current activity by client. For me, I have one channel that receives messages for a BUNCH of clients, so I want to be able to see who is outside of their historical parameters, grouped by mcm.source.

What do people do now to try to do date based queries? I'm trying to do something like this:

> select mcm.source, count(mcm.source), year(mm.received_date), month(mm.received_date), day(mm.received_date), hour(mm.received_date)
from mirth.d_mcm3 mcm
left join mirth.d_mm3 mm on mcm.message_id = mm.message_id and mm.id=0
where mm.RECEIVED_DATE > date_sub(Date(now()), interval 7 day) and mcm.METADATA_ID=0
group by mcm.source, year(mm.received_date), month(mm.received_date), day(mm.received_date), hour(mm.received_date)
order by 1,3,4,5,6

But even my staging server can't handle that. I need to use mcm.message_id, so first I need to find which IDs bracket my search range.

I see other people posting example queries that just reference the timestamp fields, but even pruning to 30 days, that just doesn't work for us.

**Reply by jonbartels** [reply]
I think in your case adding the index would be prudent. You seem to have a clear need for it to report on your routing channel. 

How is the `EXPLAIN` plan for the query you showed? You're already reducing the data set by filtering on the connector ID for source messages. At a glance, it is very driven by the `received_date` column.

Perhaps a partial index would help? Even then you still need an index and to build the partial indexes you're still going to end up with a table scan so you might as well just build an index.

**Reply by BuddhaBuddy1** [reply]
I'm not a DB guru, but still looks bad to me. I could be reading it wrong, but it looks like a 1/3 table scan? I still worry that putting an index on this will kill our DB performance. It is already a bottleneck.
<HTML>
<BODY>
<TABLE><TR><TD>1</TD><TD>SIMPLE</TD><TD>mm</TD><TD></TD><TD>ref</TD><TD>PRIMARY, D_MM14_INDEX1, D_MM14_INDEX2, D_MM14_INDEX3, D_MM14_FKI</TD><TD>D_MM14_INDEX3</TD><TD>4</TD><TD>const</TD><TD>324835</TD><TD>33.33</TD><TD>Using&nbsp;index&nbsp;condition; Using&nbsp;where; Using&nbsp;temporary; Using&nbsp;filesort</div></TD></TR><TR><TD>1</TD><TD>SIMPLE</TD><TD>mcm</TD><TD></TD><TD>eq_ref</TD><TD>PRIMARY</TD><TD>PRIMARY</TD><TD>12</TD><TD>mirth.mm.MESSAGE_ID,const</TD><TD>1</TD><TD>100.00</TD><TD></TD></TR></TABLE>

</BODY>
</HTML>

**Reply by jonbartels** [reply]
I'm not sure how to read that EXPLAIN plan from MySQL. I am a postgres fanboy, ha ha. What are the `D_MM14_INDEX##` indexes defined as? Are those indexes you added or ones generated by Mirth Connect?

You might hop on Slack https://mirthconnect.herokuapp.com/ and have a conversation with the MySQL experts that use MC and lurk in chat.

**Reply by tonygermano** [reply]
Every index you add will have a negative impact on write performance, but it should still be relatively small if you only add one or two.

One of the benefits of channels having their own tables is that the "remove all messages" method becomes a very fast table truncate rather than `delete from blah where channelid = whatever`. This has the effect of immediately freeing up disk space in the OS for most if not all db engines.

Also message Ids end up being sequential within each channel because they are not sharing a sequence.
